core/types/block_ext_test.go

```diff
@@ -0,0 +1,190 @@
+// (c) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package types
+
+import (
+	"math/big"
+	"reflect"
+	"testing"
+	"unicode"
+	"unsafe"
+
+	"github.com/ava-labs/libevm/common"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCopyHeader(t *testing.T) {
+	t.Run("empty_header", func(t *testing.T) {
+		empty := &Header{}
+
+		headerExtra := &HeaderExtra{}
+		extras.Header.Set(empty, headerExtra)
+
+		cpy := CopyHeader(empty)
+
+		want := &Header{
+			Difficulty: new(big.Int),
+			Number:     new(big.Int),
+		}
+
+		headerExtra = &HeaderExtra{}
+		extras.Header.Set(want, headerExtra)
+
+		assert.Equal(t, want, cpy)
+	})
+
+	t.Run("filled_header", func(t *testing.T) {
+		h := &Header{
+			ParentHash:       common.Hash{1},
+			UncleHash:        common.Hash{2},
+			Coinbase:         common.Address{3},
+			Root:             common.Hash{4},
+			TxHash:           common.Hash{5},
+			ReceiptHash:      common.Hash{6},
+			Bloom:            Bloom{7},
+			Difficulty:       big.NewInt(8),
+			Number:           big.NewInt(9),
+			GasLimit:         10,
+			GasUsed:          11,
+			Time:             12,
+			Extra:            []byte{13},
+			MixDigest:        common.Hash{14},
+			Nonce:            BlockNonce{15},
+			BaseFee:          big.NewInt(16),
+			WithdrawalsHash:  &common.Hash{17},
+			BlobGasUsed:      ptrTo(uint64(18)),
+			ExcessBlobGas:    ptrTo(uint64(19)),
+			ParentBeaconRoot: &common.Hash{20},
+		}
+		headerExtra := &HeaderExtra{
+			ExtDataHash:    common.Hash{21},
+			ExtDataGasUsed: big.NewInt(22),
+			BlockGasCost:   big.NewInt(23),
+		}
+		extras.Header.Set(h, headerExtra)
+
+		allFieldsAreSet(t, h)
+		allFieldsAreSet(t, headerExtra)
+
+		cpy := CopyHeader(h)
+
+		want := &Header{
```

Comment:
```suggestion
		got := CopyHeader(h)
		want := input
```

Now there's no way to make a mistake in the construction of `want`.

Reply:
Call me paranoid, but - especially in a copy function - the point is to not rely on the input for the wanted result in case the input gets mutated.

Reply:
No longer used with newer code, resolving this.

Reply:
> Call me paranoid, but - especially in a copy function - the point is to not rely on the input for the wanted result in case the input gets mutated.

Absolutely agree. That's why I implemented it as a function that returns a new instance of the same data.